### PR TITLE
Abuse detection system

### LINKED ISSUE #27: Abuse detection service
This means collecting PII-like data (IP, UA, X-Forwarded-For, etc) into a separate table, bucketed by hour (to intentionally lose temporal accuracy, to preserve privacy). 

When submissions are received, that table is consulted for how often that same PII-like value has been seen during the past 1, 3, 10 etc hours. Those values are then attached to the submission before it’s persisted. 

This acts as a “legitimacy score” of sorts, which can be used in later analysis. Far from perfect for that purpose, but strikes a balance between that usefulness and privacy of user data. 

The technical implementation for this is still undecided. Redis or some such comes to mind.

### PR DESCRIPTION
**Pull request by jareware**
This has been waiting on the backlog for a long time.

tl;dr:

* When a request comes in, we grab its source IP, its `User-Agent` and `X-Forwarded-For`, and call that its "fingerprint"
* Immediately after, we put those values through a one-way hash function, with a secret pepper included, so it can't be reversed
* For each such hashed fingerprint key/value pair, we upsert an item in DynamoDB, e.g. `"2020-03-31T10Z/source_ip/6NehvvOtv6bGjk7FTtutfcozHqX478AVLV1EbcpmV+g=" => 3`
* This means this IP has been seen 3 times within the hour-long time bucket at `2020-03-31T10:xx:yy`
* Each key has a TTL value set up in DynamoDB, so they expire 24 hours after last being written (so the DB doesn't balloon forever)
* We then calculate an "abuse score" for the incoming request, by simply tallying these values over the past 24 hours; e.g. `{ source_ip: 2, user_agent: 5, forwarded_for: -1 }` means we've seen this same IP 2 times during the past 24 hours, the `User-Agent` 5 times, and the `-1` is a special value meaning we didn't get an `X-Forwarded-For` value for this request (so it's pointless to score how common it's recently been)
* This abuse score is stored along with the response
* Later, this score can be used to assess the credibility of each response, during the data dump generation phase

Closes #27

### CODE REVIEW

File: src/backend/abuseDetection.ts [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+          RequestItems: {
+            [tableName]: {
+              AttributesToGet: ['ADKey', 'ADVal'],
+              Keys: keys.map(key => ({ ADKey: { S: key } })),
+            },
+          },
+        })
+        .promise()
+        .then(res =>
+          (res?.Responses?.[tableName] || []).reduce(
```
Comment (cutepig) [inline src/backend/abuseDetection.ts:101]: Taking the Elvis operator for a ride 🕺🏻, nice pattern here!

File: src/backend/abuseDetection.ts [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+            (memo, next) => ({ ...memo, [next.ADKey.S || '']: unwrapNumber(next.ADVal) }),
```
Comment (cutepig) [inline src/backend/abuseDetection.ts:102]: Interesting choice of naming the reducer parameters to `memo` and `next`. I always used `acc` and `value`, but these somehow look more appealing 🤔

File: src/backend/abuseDetection.ts [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+  timestamp = Date.now,
+  timeRangeHours = TIME_RANGE_HOURS,
+) {
+  const now = timestamp();
+  const timeRange = getTimeRange(now, timeRangeHours);
+  const fp = mapValues(fingerprint, v => (v ? hashFunction(v) : '')); // to preserve privacy, put all values (e.g. IP, UA) through a one-way hash function before doing anything with them (keep empty strings as empty, though, since they warrant special treatment later)
+  const keysToGet = flatMap(fingerprintKeys, key => timeRange.map(ts => getStorageKey(key, fp[key], ts)));
+  const keysToIncrement = fingerprintKeys
+    .map(key => (fp[key] ? getStorageKey(key, fp[key], now) : null))
+    .filter(nonNullable);
```
Comment (cutepig) [inline src/backend/abuseDetection.ts:41]: Tricks of the trade: You can use `Boolean` to do this unless you want to explicitly filter only `null` and `undefined` values.

File: src/backend/abuseDetection.ts (hunk 1 above)
Comment (jareware) [reply src/backend/abuseDetection.ts:101]: I remember it being called the "existential operator" in Coffee, I think :joy:

File: src/backend/abuseDetection.ts (hunk 2 above)
Comment (jareware) [reply src/backend/abuseDetection.ts:102]: Not sure where I picked that up, just happened to get etched into my brain so now I write it out automatically, otherwise no chance to remember the order of args. :stuck_out_tongue:

File: src/backend/abuseDetection.ts (hunk 3 above)
Comment (jareware) [reply src/backend/abuseDetection.ts:41]: Oh, interesting!

Though after trying this, `const keysToIncrement` is inferred as `(string | null)[]` whereas it's supposed to be just `string[]`. :thinking:

**Comment by jareware** [comment]
The utilization should be fairly simple, though admittedly somewhat fuzzy (which is a given when subject to such stringent privacy-preservation requirements): for instance, drop every response coming from an IP that has been seen `> n` times in its preceding 24 hour window. Adjusting `n` is effectively an anomaly detection exercise. The important part being that we record this information so that analysing it later is _possible,_ even if we don't make use of it immediately.

File: src/backend/abuseDetection.ts (hunk 1 above)
Comment (cutepig) [reply src/backend/abuseDetection.ts:101]: Oh, I remember Coffeescript! And with fondness too!

File: src/backend/abuseDetection.ts (hunk 3 above)
Comment (cutepig) [reply src/backend/abuseDetection.ts:41]: That's strange, I tried that on Typescript Playground to assert my claim and got `string[]` out of the `.filter` 🧐
